test/functional/hooks.js

```diff
@@ -42,37 +43,126 @@ describe('library hooks', () => {
+        stub = sinon.stub();
+        mutators.attributes.dynamicName = stub;
+      });
+
+      afterEach(() => {
+        mutators.attributes.dynamicName = null;
+      });
+
+      it('should be called for dynamic attributes', () => {
+        patch(container, render, 'dynamicValue');
+        var el = container.childNodes[0];
+
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValue');
+      });
+
+      it('should be called on attribute update', () => {
+        patch(container, render, 'dynamicValueOne');
+        patch(container, render, 'dynamicValueTwo');
+        var el = container.childNodes[0];
+
+        expect(stub).calledTwice;
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValueTwo');
+      });
 
-      expect(attributes.applyAttr).calledWith(el, 'dynamicName', 'dynamicValueTwo');
+      it('should allow only be called when attributes change', () => {
+        patch(container, render, 'dynamicValue');
+        patch(container, render, 'dynamicValue');
+        var el = container.childNodes[0];
+
+        expect(stub).calledOnce;
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValue');
+      });
+
+      it('should prioritize specific setter over generic', () => {
+        var allStub = sinon.stub();
+        mutators.attributes.all = allStub;
+        mutators.attributes.dynamicValue = stub;
+
+        patch(container, render, 'dynamicValue');
+
+        expect(stub).calledOnce;
+        expect(allStub).notCalled;
+
+        mutators.attributes.all = null;
+        mutators.attributes.dynamicValue = null;
+      });
     });
 
-    it('should allow only be called when attributes change', () => {
-      patch(container, render, 'dynamicValue');
-      patch(container, render, 'dynamicValue');
+    describe('for generic dynamic attributes', () => {
+      var allStub;
+      beforeEach(() => {
+        stub = sinon.stub();
+        mutators.attributes.all = stub;
+      });
+
+      afterEach(() => {
+        mutators.attributes.all = null;
+      });
+
+      it('should be called for dynamic attributes', () => {
+        patch(container, render, 'dynamicValue');
+        var el = container.childNodes[0];
+
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValue');
+      });
+
+      it('should be called on attribute update', () => {
+        patch(container, render, 'dynamicValueOne');
+        patch(container, render, 'dynamicValueTwo');
+        var el = container.childNodes[0];
+
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValueTwo');
+      });
+
+      it('should allow only be called when attributes change', () => {
+        patch(container, render, 'dynamicValue');
+        patch(container, render, 'dynamicValue');
+        var el = container.childNodes[0];
 
-      // Called once for the static attribute and once for the dynamic one
-      expect(attributes.applyAttr).calledTwice;
+        expect(stub).calledWith(el, 'dynamicName', 'dynamicValue');
```

Comment:
Ah, I missed the [`calledOnce`](https://github.com/google/incremental-dom/pull/92/files#diff-cd7507808e5cf41a0923565b24b06e33R114).